File: C++/Visual Studio/NetworkedSpreadsheet/PS2/DependencyGraph/DependencyGraph.cs
///<author>
///Andrew Thompson u0879848
///</author>

// Skeleton implementation written by Joe Zachary for CS 3500, September 2013.
// Version 1.1 (Fixed error in comment for RemoveDependency.)
// Version 1.2 - Daniel Kopta 
//               (Clarified meaning of dependent and dependee.)
//               (Clarified names in solution/project structure.)

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SpreadsheetUtilities
{

    /// <summary>
    /// (s1,t1) is an ordered pair of strings
    /// t1 depends on s1; s1 must be evaluated before t1
    /// 
    /// A DependencyGraph can be modeled as a set of ordered pairs of strings.  Two ordered pairs
    /// (s1,t1) and (s2,t2) are considered equal if and only if s1 equals s2 and t1 equals t2.
    /// Recall that sets never contain duplicates.  If an attempt is made to add an element to a 
    /// set, and the element is already in the set, the set remains unchanged.
    /// 
    /// Given a DependencyGraph DG:
    /// 
    ///    (1) If s is a string, the set of all strings t such that (s,t) is in DG is called dependents(s).
    ///        (The set of things that depend on s)    
    ///        
    ///    (2) If s is a string, the set of all strings t such that (t,s) is in DG is called dependees(s).
    ///        (The set of things that s depends on) 
    //
    // For example, suppose DG = {("a", "b"), ("a", "c"), ("b", "d"), ("d", "d")}
    //     dependents("a") = {"b", "c"}
    //     dependents("b") = {"d"}
    //     dependents("c") = {}
    //     dependents("d") = {"d"}
    //     dependees("a") = {}
    //     dependees("b") = {"a"}
    //     dependees("c") = {"a"}
    //     dependees("d") = {"b", "d"}
    /// </summary>
    public class DependencyGraph
    {
        private Dictionary<string, HashSet<string>> dependents; //Holds the list of dependents for a given string (the key)
        private Dictionary<string, HashSet<string>> dependees; //Holds the list of dependees for a given string (the key)
        private int size;

        /// <summary>
        /// Creates an empty DependencyGraph.
        /// </summary>
        public DependencyGraph()
        {
            dependents = new Dictionary<string, HashSet<string>>();
            dependees = new Dictionary<string, HashSet<string>>();
            size = 0;
        }


        /// <summary>
        /// The number of ordered pairs in the DependencyGraph.
        /// </summary>
        public int Size
        {
            get { return size; }
        }


        /// <summary>
        /// The size of dependees(s).
        /// This property is an example of an indexer.  If dg is a DependencyGraph, you would
        /// invoke it like this:
        /// dg["a"]
        /// It should return the size of dependees("a")
        /// </summary>
        public int this[string s]
        {
            get
            {
                //Returns 0 if the key has not been added yet.
                if(dependees.TryGetValue(s, out HashSet<string> dependeesSet))
                {
                    return dependeesSet.Count;
                }

                else
                {
                    return 0;
                }
            }
        }


        /// <summary>
        /// Reports whether dependents(s) is non-empty.
        /// </summary>
        public bool HasDependents(string s)
        {
            //Returns false if the key has not been added yet.
            if (dependents.TryGetValue(s, out HashSet<string> dependentsSet))
            {
                return dependentsSet.Count > 0;
            }

            else
            {
                return false;
            }
        }


        /// <summary>
        /// Reports whether dependees(s) is non-empty.
        /// </summary>
        public bool HasDependees(string s)
        {
            if (dependees.TryGetValue(s, out HashSet<string> dependeesSet))
            {
                return dependeesSet.Count > 0;
            }

            else
            {
                return false;
            }
        }


        /// <summary>
        /// Enumerates dependents(s).
        /// </summary>
        public IEnumerable<string> GetDependents(string s)
        {
            //Returns an empty HashSet if the key has not been added yet.
            if (dependents.TryGetValue(s, out HashSet<string> dependentsSet))
            {
                return dependentsSet.ToArray(); //Returns an array copy of the HashSet, to ensure data protection.
            }

            else
            {
                return new HashSet<string>();
            }
        }

        /// <summary>
        /// Enumerates dependees(s).
        /// </summary>
        public IEnumerable<string> GetDependees(string s)
        {
            if (dependees.TryGetValue(s, out HashSet<string> dependeesSet))
            {
                return dependeesSet.ToArray();
            }

            else
            {
                return new HashSet<string>();
            }
        }


        /// <summary>
        /// <para>Adds the ordered pair (s,t), if it doesn't exist</para>
        /// 
        /// <para>This should be thought of as:</para>   
        /// 
        ///   t depends on s
        ///
        /// </summary>
        /// <param name="s"> s must be evaluated first. T depends on S</param>
        /// <param name="t"> t cannot be evaluated until s is</param>        /// 
        public void AddDependency(string s, string t)
        {
            //Creates a key-value pair if either of the strings have not been added yet.
            if (!dependents.ContainsKey(s))
                dependents.Add(s, new HashSet<string>());

            if (!dependees.ContainsKey(t))
                dependees.Add(t, new HashSet<string>());

            //Adds the ordered pair only if it does not already exist in the HashSets yet.
            if (!dependents[s].Contains(t) && !dependees[t].Contains(s))
            {
                dependents[s].Add(t);
                dependees[t].Add(s);
                size++;
            }

            else
            {
                return;
            }
        }


        /// <summary>
        /// Removes the ordered pair (s,t), if it exists
        /// </summary>
        /// <param name="s"></param>
        /// <param name="t"></param>
        public void RemoveDependency(string s, string t)
        {
            //Only goes to remove the pair if both keys exist and the strings exist in the HashSets.
            if (dependents.ContainsKey(s) && dependees.ContainsKey(t))
            {
                if (dependents[s].Contains(t) && dependees[t].Contains(s))
                {
                    dependents[s].Remove(t);
                    dependees[t].Remove(s);
                    size--;
                }

                else
                {
                    return;
                }
            }

            else
            {
                return;
            }
        }


        /// <summary>
        /// Removes all existing ordered pairs of the form (s,r).  Then, for each
        /// t in newDependents, adds the ordered pair (s,t).
        /// </summary>
        public void ReplaceDependents(string s, IEnumerable<string> newDependents)
        {
            //Here all ordered pairs (s,r) are removed, if any.
            if (dependents.TryGetValue(s, out HashSet<string> dependentsSet))
            {
                foreach (string r in GetDependents(s))
                {
                    RemoveDependency(s, r);
                }
            }

            //Here, the new ordered pairs are added.
            foreach(string t in newDependents)
            {
                AddDependency(s, t);
            }
        }


        /// <summary>
        /// Removes all existing ordered pairs of the form (r,s).  Then, for each 
        /// t in newDependees, adds the ordered pair (t,s).
        /// </summary>
        public void ReplaceDependees(string s, IEnumerable<string> newDependees)
        {
            if (dependees.TryGetValue(s, out HashSet<string> dependeesSet))
            {
                foreach (string r in GetDependees(s))
                {
                    RemoveDependency(r, s);
                }
            }

            foreach (string t in newDependees)
            {
                AddDependency(t, s);
            }
        }

    }

}

File: C++/Visual Studio/NetworkedSpreadsheet/dependency_graph.cpp
#include "dependency_graph.h"

#include <algorithm>

using namespace std;

dependency_graph::dependency_graph(){
  this->dependents = map<string, vector<string> >();
  this->dependees = map<string, vector<string> >();

  this->size = 0;
}

int dependency_graph::get_size(){
  return this->size;
}

int dependency_graph::get_num_dependees(string s){
  if(this->dependees.find(s) == this->dependees.end()){
    return 0;
  }
  else{
    return this->dependees[s].size();
  }
}

bool dependency_graph::has_dependents(string s){
  if(this->dependents.find(s) == this->dependents.end()){
    return false;
  }
  else{
    return this->dependents[s].size() > 0;
  }
}

bool dependency_graph::has_dependees(string s){
  if(this->dependees.find(s) == this->dependees.end()){
    return false;
  }
  else{
    return this->dependees[s].size() > 0;
  }
}

vector<string> dependency_graph::get_dependents(string s){
  if(this->dependents.find(s) == this->dependents.end()){
    return vector<string>();
  }
  else{
    return this->dependents[s];
  }
}

vector<string> dependency_graph::get_dependees(string s){
  if(this->dependees.find(s) == this->dependees.end()){
    return vector<string>();
  }
  else{
    return this->dependees[s];
  }
}

void dependency_graph::add_dependency(string s, string t){
  if(this->dependents.find(s) == this->dependents.end()){
    this->dependents.insert(pair<string, vector<string> >(s, vector<string>()));
  }

  if(this->dependees.find(t) == this->dependees.end()){
    this->dependees.insert(pair<string, vector<string> >(t, vector<string>()));
  }

  if(find(this->dependents[s].begin(), this->dependents[s].end(), t) == this->dependents[s].end() &&
     find(this->dependees[t].begin(), this->dependees[t].end(), s) == this->dependees[t].end()){
    this->dependents[s].push_back(t);
    this->dependees[t].push_back(s);
    this->size++;
  }
}

void dependency_graph::remove_dependency(string s, string t){
  if(this->dependents.find(s) == this->dependents.end() && 
     this->dependees.find(s) == this->dependees.end()){
    return;
  }

  else{
    if(find(this->dependents[s].begin(), this->dependents[s].end(), t) == this->dependents[s].end() &&
       find(this->dependees[t].begin(), this->dependees[t].end(), s) == this->dependees[t].end()){
      return;
    }

    else{
      this->dependents[s].erase(find(this->dependents[s].begin(), this->dependents[s].end(), t));
      this->dependees[t].erase(find(this->dependees[t].begin(), this->dependees[t].end(), s));
    }
  }
}

void dependency_graph::replace_dependents(string s, vector<string> new_dependents){
  if (this->dependents.find(s) != this->dependents.end()) {
    vector<string> old_dependents = this->get_dependents(s);
    for (int i = 0; i < old_dependents.size(); i++) {
      this->remove_dependency(s, old_dependents[i]);
    }
  }
  for (int i = 0; i < new_dependents.size(); i++) {
      this->add_dependency(s, new_dependents[i]);
    }
}

void dependency_graph::replace_dependees(string s, vector<string> new_dependees){
  if (this->dependees.find(s) != this->dependees.end()) {
    vector<string> old_dependees = this->get_dependees(s);
    for (int i = 0; i < old_dependees.size(); i++) {
      this->remove_dependency(old_dependees[i], s);
    }
  }
  for (int i = 0; i < new_dependees.size(); i++) {
    this->add_dependency(new_dependees[i], s);
  }
}

bool dependency_graph::check_circular(string s, vector<string> direct_dependencies) {

  vector<string> old_dependees = this->get_dependees(s);
  this->replace_dependees(s, direct_dependencies);

  vector<string> visited;
  
  //getcellstorecalculate
  try {
    for (int i = 0; i < direct_dependencies.size(); i++) {
      if (find(visited.begin(), visited.end(), direct_dependencies[i]) == visited.end()) {
        this->visit(direct_dependencies[i], direct_dependencies[i], visited);
      }
    }
    return false;
  }
  catch(int e) {
    this->replace_dependees(s, old_dependees);
    return true;
  }
}

void dependency_graph::visit(string start, string name, vector<string> &visited) {
  vector<string> sub_dependents = this->get_dependents(name);
  visited.push_back(name);
  for (int i = 0; i < sub_dependents.size(); i++) {
    if (sub_dependents[i] == start) {
      throw 2;
    }
    else if (find(visited.begin(), visited.end(), sub_dependents[i]) == visited.end()) {
      this->visit(start, sub_dependents[i], visited);
    }
  }
}
